### PR TITLE
Support the new cover format

### LINKED ISSUE #54: Any plans to support the upcoming library UI change for game boxart?
https://www.tweaktown.com/news/66182/steams-new-ui-leaks-look-great/index.html

Steam is switching to standard boxart game images. Naturally, it's going to take ages to make these changes by hand.

It would be awesome if SteamGrid was updated to grab boxart instead of the current 460x215 images.

Thanks for all your hard work @boppreh

### PR DESCRIPTION
**Pull request by Lucki**
This adds support for the new library format (600x900).
Google search is disabled because of the really bad results. We should add some additional sources like IGDB or SteamGridDB but this will need an api key as user input/parameter/config.
Apng gets preserved if no overlay is applied.

I've never done something in GO before so let me know if there's something to improve.

Fix #54

### CODE REVIEW

**Comment by Lucki** [comment]
SteamGridDBs api doesn't have a filter for the new format yet so right now it takes the top voted one which may not be one in the new format.

**Comment by Lucki** [comment]
SteamGridDB has implemented the required api calls. Since this still hasn't been merged or received comments I have merged SteamGridDB and IGDB support into this PR.

You can optionally provide your api keys ([SteamGridDB](https://www.steamgriddb.com/profile/preferences), [IGDB](https://api.igdb.com/signup)) like this:
~~~
./steamgrid --steamgriddb "steamgriddb-api-key" --igdb "igdb-api-key"
~~~

**Comment by strzlee** [comment]
This doesn't work for "Non-Steam-Games". Seems the new cover-files for non-steam-games are named different than the old ones. 

Example (GOG Version of the Witcher 3):
old grid image: 13431636140889210880.jpeg
new cover image : 3127296488p.jpg

**Comment by Lucki** [comment]
Steam refused to set a custom image for my added games.
Do they work for you when set from inside steam? (I guess this is a yes)
Do you know how the new name is known? The current/old way is here: https://github.com/boppreh/steamgrid/blob/eac4d3fef11a1f4bf6f85d1e6fc0ff089c5fcbd1/games.go#L114-L120

**Comment by boppreh** [comment]
Call me crazy, but I still cannot find this new cover format used anywhere on Steam. Is this is an Open Beta UI kind of thing?

**Comment by PoorPocketsMcNewHold** [comment]
> Call me crazy, but I still cannot find this new cover format used anywhere on Steam. Is this is an Open Beta UI kind of thing?

It is. Go opt-in for the Steam Open Beta to test it out. 
Capture d'écran Deepin_20191020153045

**Comment by strzlee** [comment]
> Steam refused to set a custom image for my added games. (nothing happens when I set a custom artwork)
> Do they work for you when set from inside steam? (I guess this is a yes)
> Do you know how the new name is known? The current/old way is here:
> 
> https://github.com/boppreh/steamgrid/blob/eac4d3fef11a1f4bf6f85d1e6fc0ff089c5fcbd1/games.go#L114-L120
> 
> When comparing with my file I spotted some differences in my actual file:
> 
> ```diff
> - (?i)appname\x00(.+?)\x00\x01exe\x00(.+?)\x00\x01.+?\x00tags\x00(.*?)\x08\x08
> + (?i)\x00\x01appname\x00(.+?)\x00\x01exe\x00(.+?)\x00\x01.+?\x00tags\x00\x01(.*?)\x08\x08
> ```

The filename for the new ui covers is what the operation at line 119 generates.

For my example "The Witcher 3 - Wild Hunt":

fmt.Println(top)

> 3127296488

fmt.Println(gameID)

> 13431636140889210880

image-files in ".../config/grid/":

3127296488p.jpg 
13431636140889210880.jpeg

**Comment by Lucki** [comment]
Thanks @strzlee! While I'm still unable to set them manually through steam this worked:

grafik

**Comment by strzlee** [comment]
> Setting the hero image and the logo is now possible too, yet I'm still unable to set them here so I'd appreciate if someone can tell me if the naming scheme is somehow similar.

According to my previous example:

```
Cover: 3127296488p.jpg
Hero: 3127296488_hero.jpg
Grid: 13431636140889210880.jpg
```

**Comment by Lucki** [comment]
Are you sure about that grid image?
The grid image I see is set through `3127296488.jpg` (without an additional extension). It's the first shown image when hovering over a cover or can be seen on the last played slot.
Can you also look into the logo naming?

**Comment by strzlee** [comment]
With "Grid" I refer to the classic format for "non-Beta-Clients" respectively the "Big-Picture-Mode".

**Comment by Lucki** [comment]
@boppreh The new library is now live for everyone - no need to opt into a beta.

**Comment by strzlee** [comment]
Something in the download/search algorithm is subpar compared with the latest release version from boppreh.

Lucki:new_cover branch downloads often as banners some in-game-screenshots with wrong dimensions where boppreh:master delivers always the correct artwork. Especially for non-steam-games / shortcuts.

**Comment by Lucki** [comment]
Thanks for testing @strzlee. That was actually an oversight in the latest commit. The google seach should now search for the legacy grid image dimensions again (460 x 215). The format was mostly common for steam grids hence the google search was mostly successful but is garbage on the other formats.

**Comment by strzlee** [comment]
Because i couldn't open an issue at the page of your fork, here another little bug:

https://github.com/Lucki/steamgrid/blob/ba56c7fba2a6a07d58d2f40982c91b22dbaed144/download.go#L164-L178

You compare "jsonSearchResponse.Data[0].Types" (type SteamGridDBSearchResponse) with "artTypes". But the "type" param from steamgriddb-api response to "/search/autocomplete/" don't refers to the "style". "Type" holds an array with the different stores.

An example: 

```
api-call: https://www.steamgriddb.com/api/v2/search/autocomplete/Watch_Dogs?styles=alternate&dimensions=920x430

response-body: 
{"success":true,"data":
[
   {"id":2823,"name":"Watch_Dogs","types":["steam","origin","uplay"],"verified":true}, 
   {"id":10872,"name":"Watch_Dogs 2","types":["egs","steam","origin","uplay"],"verified":true}, 
   {"id":11321,"name":"Watch This!","types":["steam"],"verified":true}
]} 
```

**Comment by strzlee** [comment]
> Thanks for testing @strzlee. That was actually an oversight in the latest commit. The google seach should now search for the legacy grid image dimensions again (460 x 215). The format was mostly common for steam grids hence the google search was mostly successful but is garbage on the other formats.

Doesn't work like expected, but found the issue. Sprintf returns a malformed url:

> https://www.google.com.br/search?tbs=isz%!A(string=460)ex%!C(string=215)iszw%!A(MISSING)%!v(MISSING)%!C(MISSING)iszh%!A(MISSING)%!v(MISSING)&tbm=isch&num=5&q=

This string should work:

`const googleSearchFormat = 'https://www.google.com.br/search?tbs=isz%%3Aex%%2Ciszw%%3A%v%%2Ciszh%%3A%v&tbm=isch&num=5&q='`

 (I'm to lazy for pushing a commit and opening pull request :p)

**Comment by Lucki** [comment]
Thanks again! SteamGridDB referenced these as `platform` everywhere else so I think i guessed what `type` is and forgot to check.

**Comment by strzlee** [comment]
Works for me so far @Lucki

steamgrid-test

**Comment by Lucki** [comment]
Thanks again for testing!

I'll wait for steamgriddb to add support for logos and heroes and after a bit of testing I will provide a precompiled binary until this gets merged.

**Comment by Logerfo** [comment]
@Lucki does it also work for animated covers or this would still need some extra work?

**Comment by Lucki** [comment]
Animated covers should work as long as no overlay is applied on top.

**Comment by strzlee** [comment]
> Thanks again for testing!
> 
> I'll wait for steamgriddb to add support for logos and heroes and after a bit of testing I will provide a precompiled binary until this gets merged.

I uploaded a test-build which includes Lucki's latest patches and a experimental configuration gui here:

[https://github.com/strzlee/steamgrid/releases](https://github.com/strzlee/steamgrid/releases)

**Comment by Lucki** [comment]
Apngs are now supported but there are still some issues I have to figure out while sticking the overlay on top.
Also: They're getting HUGE.

**Comment by Lucki** [comment]
@Logerfo APNGs are working now. I advise to not use an overlay with something half transparent as that'll get darker over time.

**Comment by strzlee** [comment]
@Lucki 

Sometimes I encounter timeouts and errors from the SteamGridDB-API requests. Maybe some sort of rate-limiting?! Reducing (or slowing down) the calls would perhaps help. 

E.g.: For non-steam-games you call a GET request with the steam ID (which always fails), 
then a GET /search/game-name & a GET to /game/SteamGridID. The first request, which always fails, could here be avoided.

**Comment by Logerfo** [comment]
@Lucki I think it would be very useful to have a filter that only looks for animated covers. I want animated covers for my games, but I have a lot of games in my library, so it would take a very long time to search for animated covers for all of them.

**Comment by Lucki** [comment]
@Logerfo `--help` lists all available options

**Comment by boppreh** [comment]
@Lucki, what's the current status of this PR? Is it ready to merge, do you want to add more stuff, do you need help fixing anything?

**Comment by Lucki** [comment]
@boppreh 
I wanted to add logo download from steamgriddb but that doesn't seems to be that near:
> [I don't feel like touching the site until the Manager is fixed again.](https://discordapp.com/channels/488621078302949377/552548116700594176/644736661548433408)

So I think this is ready? Couldn't find any problem in my test runs. There aren't any remaining problems mentioned from @strzlee and @PoorPocketsMcNewHold , although I'm unsure about the timeout issue. While a lot of options are added the fire and forget design remained intact.
Since there wasn't much feedback I'm unsure if it's **good** but at least it'll work.

**Comment by strzlee** [comment]
> @boppreh
> I wanted to add logo download from steamgriddb but that doesn't seems to be that near:
> 
> > [I don't feel like touching the site until the Manager is fixed again.](https://discordapp.com/channels/488621078302949377/552548116700594176/644736661548433408)
> 
> So I think this is ready? Couldn't find any problem in my test runs. There aren't any remaining problems mentioned from @strzlee and @PoorPocketsMcNewHold , although I'm unsure about the timeout issue. While a lot of options are added the fire and forget design remained intact.
> Since there wasn't much feedback I'm unsure if it's **good** but at least it'll work.

I tested the branch in his current state on several windows and linux machines with different configurations and stepped through a couple of debug sessions with "_delve_". 
From my view it's **stable** and **works** well.

I only encountered some inexplicable results when querying the steamgriddb-api for banner with the "_dimensions_" parameter. But that's probably a server-side issue.

**Comment by Lucki** [comment]
Since nothing's going on here or at steamgriddb I've uploaded prebuild binaries: https://github.com/Lucki/steamgrid/releases/tag/v3.0.0

**Comment by powei94** [comment]
> Since nothing's going on here or at steamgriddb I've uploaded prebuild binaries: https://github.com/Lucki/steamgrid/releases/tag/v3.0.0

Hey thanks for adding builds, I tried it on windows and it works but for some reason it ignores non steam games, not sure what the issue is.

**Comment by Lucki** [comment]
Please have a look into your `/userdata/<id>/config/shortcuts.vdf` with a hex editor and verify the sequence is similar to the [mentioned above](https://github.com/boppreh/steamgrid/pull/57#issuecomment-541425126):

> ```diff
> - (?i)appname\x00(.+?)\x00\x01exe\x00(.+?)\x00\x01.+?\x00tags\x00(.*?)\x08\x08
> + (?i)\x00\x01appname\x00(.+?)\x00\x01exe\x00(.+?)\x00\x01.+?\x00tags\x00\x01(.*?)\x08\x08
> ```

**Comment by Perturbator666** [comment]
> Since nothing's going on here or at steamgriddb I've uploaded prebuild binaries: https://github.com/Lucki/steamgrid/releases/tag/v3.0.0

Can't locate STEAMPATH

**Comment by Lucki** [comment]
I haven't changed anything substantial in the steam detection. Did it work for you with the old version?
Please try again by giving the steam path with `--steamdir`.

**Comment by Perturbator666** [comment]
> I haven't changed anything substantial in the steam detection. Did it work for you with the old version?
> Please try again by giving the steam path with `--steamdir`.

That worked.

File: steamgrid.go [hunk 1]
```diff
@@ -26,8 +28,38 @@ func main() {
...
+		// BannerLQ: 460 x 215
+		// BannerHQ: 920 x 430
+		// CoverLQ: 300 x 450
+		// CoverHQ: 600 x 900
+		// HeroLQ: 1920 x 620
+		// HeroHQ: 3840 x 1240
+		// LogoLQ: 640 x 360
+		// LogoHQ: 1280 x 720
+		// artStyle: ["idExtension", "nameExtension", steamExtension, dimXHQ, dimYHQ, dimXLQ, dimYLQ]
+		"Banner": []string{"", ".banner", "header.jpg", "920", "430", "460", "215"},
```
Comment (boppreh) [inline steamgrid.go]: Should probably use a record/struct value instead of an array of strings, so that accesses are clearer.

File: README.md [hunk 2]
```diff
@@ -16,10 +16,23 @@ again when you get more games or want to update the category overlays.
...
+    * Add the extension `.banner` before the image extension for banner art: `games i love.banner.png`
+    * Add the extension `.cover` before the image extension for cover art: `games i love.cover.png`
+    * Add the extension `.hero` before the image extension for hero art `games i love.hero.png`
+    * Add the extension `.logo` before the image extension for logo art `games i love.logo.png`
+3. *(optional)* Download a pack of custom images and place it in the `games/` folder. The image files can be either the name of the game (e.g. `Psychonauts.banner.png`) or the game id (e.g. `3830.png`).
+    * Add the extension `.banner` before the image extension for banner art: `Psychonauts.banner.png`, `3830.png`
+    * Add the extension `.cover`/`p` before the image extension for cover art: `Psychonauts.cover.png`, `3830p.png`
+    * Add the extension `.hero`/`_hero` before the image extension for hero art `Psychonauts.hero.png`, `3830_hero.png`
+    * Add the extension `.logo`/`_hero` before the image extension for logo art `Psychonauts.logo.png`, `3830_logo.png`
+4. *(optional)* Generate a some API Keys to enhance the automatic search:
```
Comment (boppreh) [inline README.md]:
```suggestion
4. *(optional)* Generate some API Keys to enhance the automatic search:
```

**Comment by boppreh** [comment]
Looks stable enough and also passed my tests.

That was an amazing tour de force. Thank you very much for this contribution, from the large parts (cover format, new APIs) to the small (more reliable parsing of steam files).

**Comment by boppreh** [comment]
New release! My superficial tests worked, but please let me know if you run into any problems or have further suggestions. Unfortunately I don't have a lot of spare time to dedicate to this tool, but I'm glad it's still helping people.

**Comment by gitscosh** [comment]
Thank you to everyone who made this happen!

**Comment by Lucki** [comment]
> It's a bit strange that the covers are cut off at the bottom. Here you can see my test overlay where the bottom outline isn't visible: 
<details>
<summary>Picture</summary>

> grafik

</details>

Today's Steam Beta update fixed this issue:
> [Fix game images being truncated by a couple of pixels in shelves and collections.](https://steamcommunity.com/groups/SteamClientBeta#announcements/detail/3114798907236840432)

**Pull request merged**